Feature/hint compat api css

CKGrafico opened the pull request:
---------
This PR is in development, check first comment
---------
This PR has to implement 2 of the multiple hints about Compat Api everything is detailed in https://github.com/webhintio/rfcs/issues/10

But I'm moving the checklist to this PR:
To do for CSS part of this task:
* [x] Create: Multi hint named `hint-compat-api` and organize it by language (CSS, JS, HTML)
* [x] Create: Also split each of this by deprecated and not broadly supported
* [X] Feature: Which properties or values are deprecated in the CSS
    * [x] To Do: Take Browsers data from config.
    * [x] To Do: Take MDN data for the API to check (CSS) (Generic service).
    * [x] To Do: Merge this two sources to have one source of truth.
    * [x] To Do: For Each parsed key (.walk()) check if is suported or not.
    * [x] To Do: Report if there is an error
    * [x] To Do:  Add tests
    * [x] To Do:  Docs
* [x] Feature: Which properties or values are not broadly supported in the CSS.
    * [X] To Do: For Each parsed key (.walk()) check if is not broadly supported.
    * [X] To Do: Report if there is an error
    * [X] To Do: Add tests
    * [x]  To Do: Docs

* [x] Feature: Sometimes features are multilevel and they should work.
* [x] Feature: If the user had user prefixes for one browser and feature has to not report erros in not broadly supported properties.

* [x] Check: Avoid to throw errors if the user configured to do not support a specified browser (ex: API is deprecated for IE but the user configured hint only for Chrome) (Maybe we need more info to be sure about this point)

* [x] To Do: Use Location into reports
* [x] To Do: Don't check more than once the same scenario (cache)
* [x] To Do: Check foreach loops

* [x] Feedback: Add location problem to tests

* [x] Review: All relevant code is documented
* [x] Review: All the functionality is covered by tests and is near 100% of coverage
* [x] Review: All the functionality is documented in docs
* [x] Review: Check with some real websites

Bugs or questions for 3rd party:
* [x] https://github.com/mdn/browser-compat-data/issues/3034
* [x] https://github.com/webhintio/rfcs/issues/10#issuecomment-433424902

CKGrafico commented:
This PR is in development but feel free to give feedback about this files:
 - 
 -
More info in https://github.com/webhintio/rfcs/issues/10 (structure and details)

molant commented:
@antross I believe this is ready for a first pass of feedback if you have time.

antross commented on packages/hint-compat-api/src/helpers/compat-api.ts:
```diff
@@ -0,0 +1,152 @@
+        return compatDataApi;
+    }
+
+    public getPrefix(name: string): [string, string] | [undefined, string] {
```
Slightly more concise:
```suggestion
    public getPrefix(name: string): [string | undefined, string] {
```

CKGrafico commented on packages/hint-compat-api/src/compat-api-css.ts:
```diff
@@ -0,0 +1,137 @@
+        const onParseCSS = (styleParse: StyleParse): void => {
+            const { resource } = styleParse;
+            const mdnBrowsersCollection = userBrowsers.convert(context.targetedBrowsers);
+            const compatApi = new CompatApi('css', mdnBrowsersCollection);
```
You're perfectly right! everything should be outside onParse

molant commented on packages/hint-compat-api/src/helpers/compat-api.ts:
```diff
@@ -0,0 +1,156 @@
+        return matched && matched.length > 0 ? [matched[0], name.replace(matched[0], '')] : [undefined, name];
+    }
+
+    public getSupportStatementFromInfo(browserFeatureSupported?: SupportStatement, prefix?: string): SimpleSupportStatement | undefined {
```
I don't think we are memoizing the result of anything (PR is big so I might have missed it) and I imagine we will be checking multiple times the same properties. If that's the case we should look into adding that.

CKGrafico replied:
Adding as a task

molant commented on packages/hint-compat-api/src/helpers/compat-api.ts:
```diff
@@ -0,0 +1,156 @@
+        let isRequiredToTest = false;
+
+        forEach(this.browsers, (browserVersionsList, browser) => {
+            if (isRequiredToTest || !typedFeatureValue.__compat || !typedFeatureValue.__compat.support) {
```
I'd do something similar to what I proposed in my previous comment about using `every` to validate everything so it's easier to know what we are validating.

molant commented on packages/hint-compat-api/src/helpers/compat-api.ts:
```diff
@@ -0,0 +1,156 @@
+
+                if (addedVersion || browserVersionsList[0] <= browserVersions.normalize(addedVersion)) {
+                    isRequiredToTest = true;
+                }
```
Maybe add a `return` here so we can remove the else in line 143 and have the next `if`s with less nesting?

molant commented on packages/hint-compat-api/src/helpers/compat-api.ts:
```diff
@@ -0,0 +1,156 @@
+        // TODO: Here we are checking only parent but this object has children
+        let isRequiredToTest = false;
+
+        forEach(this.browsers, (browserVersionsList, browser) => {
```
If I've understood the code correctly, this could be improved by doing [`Array.prototype.some()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) so the loop stops the moment there's a `true` value returned.

CKGrafico replied:
Yes! nice feedback, we will use _.some to be easier for collections if this is ok

molant replied:
I'd rather us use the builtin methods. So far the only methods used from Lodash are `forEach` and now `some` too which IMO doesn't justify using it and pulling another package.

antross commented on packages/hint-compat-api/src/compat-api-css-next.ts:
```diff
@@ -0,0 +1,162 @@
+
+export default class implements IHint {
+
+    public static readonly meta: HintMetadata = {
```
Just a heads-up: we're changing how `HintMetadata` is declared to facilitate loading it separate from hints (e.g. for building documentation). See `hint-typescript-config` in PR #1480 for how I'm proposing we do this for multi-hints.

CKGrafico replied:
Ok thanks

molant commented:
Nice work so far @CKGrafico and @isabellachen !

CKGrafico commented:
Added some improvements after check with real websites.
Added new task (To add problem location to the report)
Pending: Check @molant Feedback

CKGrafico commented on packages/hint-compat-api/src/compat-api-css-next.ts:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    public constructor(context: HintContext) {
+        const mdnBrowsersCollection = userBrowsers.convert(context.targetedBrowsers);
```
@isabellachen can you check this?

isabellachen commented:
@CKGrafico CompatApi is not taking care of "version_added": "null" for children. It currently omits any feature that is not required to test only up to one level...

isabellachen commented:
@CKGrafico Another edge case with the mdn compat data: compat data for flex properties, like justify-content: space-evenly are three levels deep as the compat data for space-evenly is stored in a 'flex_context' property under justify-content. I.e. data['justify-content']['flex_context']['space-evenly']

CKGrafico commented:
Review tasks:

* [ ] Add concrete examples in normalize functions (xx.xx.xx)
* [ ] Test that all the browser names are ok, on browser covnersions
* [ ] Add concrete examples in getWorstCaseSupportStatementFromInfo
* [ ] In compat-api.ts divide bool check and number check in two blocks
* [ ] Show one error for eac cached feature (and new problem location)
* [ ] Dblcheck all the any, maybe some of them could be replaced
* [ ] Check the `undefined`insed compat-api
* [ ] Add concrete examples in getSupportStatementFromInfo
* [ ] Be sure that is not throwing errors in scenarios like:
```
    -moz- : added 21
    '' : added 35
 
    supporting 34-35
```
* [ ] Join error versions like:
```
chrome 32, chrome 33, chrome 34
chrome 32-35
```
* [ ] Use singular if the list of errors has 1 item
* [ ] Array<HintTest>  is HintTest[]

CKGrafico commented:
@antross @molant 
Today I've discovered that there are some basic css properties like `text-transform`https://developer.mozilla.org/es/docs/Web/CSS/text-transform or `min-height` https://developer.mozilla.org/es/docs/Web/CSS/min-height 

where we are reporting errors like:

```
line 6    col 10  Error  text-transform of CSS is not added on chrome_android browser.                                                       compat-api-css-next
line 7    col 10  Error  text-transform of CSS is not added on samsunginternet_android browser.                                              compat-api-css-next
```
only on mobile browsers, this is happening because in the MDN API there is no info about this browsers.

I think that to report this error is technically correct, but maybe can confuse to the user.
Do you still think that  the best is to do PR to MDN package and try to update their API or do you prefer to do something manually for our package?

molant commented:
Do a pr to MDN. We want the data coming from there as accurate as possible.
Thanks!


________________________________
From: Quique Fdez Guerra <notifications@github.com>
Sent: Monday, November 19, 2018 01:29
To: webhintio/hint
Cc: Antón Molleda; Mention
Subject: Re: [webhintio/hint] Feature/hint compat api css (#1443)


@antross<https://github.com/antross> @molant<https://github.com/molant>
Today I've discovered that there are some basic css properties like text-transformhttps://developer.mozilla.org/es/docs/Web/CSS/text-transform or min-height https://developer.mozilla.org/es/docs/Web/CSS/min-height

where we are reporting errors like:

line 6    col 10  Error  text-transform of CSS is not added on chrome_android browser.                                                       compat-api-css-next
line 7    col 10  Error  text-transform of CSS is not added on samsunginternet_android browser.                                              compat-api-css-next


only on mobile browsers, this is happening because in the MDN API there is no info about this browsers.

I think that to report this error is technically correct, but maybe can confuse to the user.
Do you still think that the best is to do PR to MDN package and try to update their API or do you prefer to do something manually for our package?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/webhintio/hint/pull/1443#issuecomment-439826868>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAlBglFgOXKy0Yuq5XXFp3EIclU1Hiceks5uwnn5gaJpZM4X5EZt>.

CKGrafico commented:
Merging the feedback from the review @molant.

I decided to move this task apart because has more work than expected.

* [ ] Be sure that is not throwing errors in scenarios like:
```
    -moz- : added 21
    '' : added 35
 
    supporting 34-35
```

After this task the last thing will be autoprefixer ¿Do you want to use the same PR?

CKGrafico commented on packages/hint-compat-api/src/compat-api-css-next.ts:
```diff
@@ -0,0 +1,162 @@
+        const onParseCSS = (styleParse: StyleParse): void => {
+            const { resource } = styleParse;
+
+            compatCSS.searchCSSFeatures(compatApi.compatDataApi, mdnBrowsersCollection, styleParse, resource);
```
Yes

CKGrafico commented on packages/hint-compat-api/src/compat-api-css-next.ts:
```diff
@@ -0,0 +1,162 @@
+                    });
+
+                    if (!wasSupportedInSometime) {
+                        context.report(resource, null, `${featureName} of CSS was never added on ${browserToSupportName} browser.`, featureName);
```
Second time that we forget this.. solved. Maybe we need to add some other async / await but I have not found more.

molant commented:
@CKGrafico we've merged the metadata change. If you can update your PR to follow that I think this should be ready for merge and continue working on the things that you mention in your comment.

@antross, @sarvaje, @alrra thoughts?

alrra commented on packages/hint-compat-api/package.json:
```diff
@@ -0,0 +1,71 @@
+  "repository": "webhintio/hint",
+  "scripts": {
+    "build": "npm run clean && npm-run-all build:*",
+    "build-release": "npm run clean && npm run build:assets && tsc -b --inlineSourceMap false --removeComments true",
```
```suggestion
    "build-release": "npm run clean && npm run build:assets && tsc --inlineSourceMap false --removeComments true",
```

Ref: https://github.com/webhintio/hint/pull/1411

alrra commented on packages/hint-compat-api/docs/compat-api-css-next.md:
````diff
@@ -0,0 +1,121 @@
+}
+```
+
+The at-rule [keyframes](https://developer.mozilla.org/en-US/docs/Web/CSS/@keyframes) was added for versions of Chrome 43 and onwards. Using `@keyframes`
````
Keep the line length under 80 characters.

alrra commented on packages/hint-compat-api/docs/compat-api-css.md:
```diff
@@ -0,0 +1,96 @@
+
+## Can the hint be configured?
+
+This hint throws errors for CSS features that have been deprecated in any of the [targeted browsers](../../hint/docs/user-guide/configuring-webhint/browser-context.md) listed.
```
Keep the line under 80 characters.

alrra commented on packages/hint-compat-api/docs/compat-api-css.md:
```diff
@@ -0,0 +1,96 @@
+
+This hint throws errors for CSS features that have been deprecated in any of the [targeted browsers](../../hint/docs/user-guide/configuring-webhint/browser-context.md) listed.
+
+The targeted browsers can be defined in either the `.hintrc` or `package.json` file. This property follows the same convention as [browserlist](https://github.com/browserslist/browserslist#readme).
```
Keep the line under 80 characters.

sarvaje commented on packages/hint-compat-api/src/helpers/compat-api.ts:
```diff
@@ -0,0 +1,223 @@
+    }
+
+    public isBrowserToSupportPartOfBrowsersCollection(browsersToSupport: BrowserSupportCollection, browserToSupportName: string): boolean {
+        if (!Object.keys(browsersToSupport).some((browser) => {
```
Move the condition to another method, it is a little bit difficult to read.

CKGrafico replied:
I hope is better now:
```
const isBrowserInsideCollection = Object.keys(browsersToSupport).some((browser) => {
            return browser === browserToSupportName;
        })

        return isBrowserInsideCollection;
```

sarvaje commented on packages/hint-compat-api/src/helpers/compat-css.ts:
```diff
@@ -0,0 +1,285 @@
+                    grouped += `${browser} ${value}`;
+                }
+
+                if (nextValue && Number(nextValue) - Number(value) > 1) {
```
value and nextValue is not a number by default?

CKGrafico replied:
Can be something like 4.4.4 https://github.com/mdn/browser-compat-data/issues/3034
![image](https://i.imgur.com/vCsjPRy.png)

But I've updated to be safe against '4.4.4' because we forgot, thanks.

CKGrafico commented:
@antross @molant merged with master and updated metadata + events names + context.report changes.

sarvaje commented on packages/hint-compat-api/README.md:
```diff
@@ -0,0 +1,51 @@
+
+## Why is this important?
+
+Explain why this package is important for your users
```
I have talked to @molant and this is not necessary yet.

CKGrafico replied:
Solved with @alrra feedback

sarvaje commented on packages/hint-compat-api/README.md:
```diff
@@ -0,0 +1,51 @@
+
+## Further Reading
+
+What can the user read to know more about this subject?
```
I have talked to @molant and this is not necessary yet.

CKGrafico commented on packages/hint-compat-api/README.md:
```diff
@@ -0,0 +1,51 @@
+# compat-api (`@hint/hint-compat-api`)
+
+description
```
Modified, inspired in https://github.com/webhintio/hint/edit/master/packages/hint-typescript-config/README.md

CKGrafico commented:
Applied @alrra feedback about md, thanks

CKGrafico commented on packages/hint-compat-api/src/compat-api-css-next.ts:
```diff
@@ -0,0 +1,149 @@
+
+        this.mdnBrowsersCollection = userBrowsers.convert(context.targetedBrowsers);
+        this.compatApi = new CompatApi('css', this.mdnBrowsersCollection, isCheckingNotBroadlySupported);
+        this.compatCSS = new CompatCSS(context, (...params) => {
```
This is  not going to work :(

sarvaje replied:
why?

CKGrafico replied:
ok after talk we are going to change arrow functions to bind(this), because the ideal case is one line arrow functions but are not allowed by linting. @borgitas21 we can talk about this tomorrow.

We can talk in the future with @molant and @sarvaje to allow one line lambdas, but not important at this moment :)

How we understand what is better:
ok
```
this.compatCSS = new CompatCSS(context, (...params) => {
    this.testFeatureIsSupportedInBrowser(...params);
});
```

better
```
new CompatCSS(context, this.testFeatureIsSupportedInBrowser.bind(this));
```

best
```
new CompatCSS(context, (...x) => this.testFeatureIsSupportedInBrowser(...x));
```

sarvaje replied:
IMO, we should go with bind even if we enable one line lambdas.

The effect is the same, we are creating a new function in both cases, but in the case of the `bind`, you don't care about the parameters or anything else.

At least for me, `bind` is cleaner than one line lambda.

@webhintio/contributors thoughts?

borgitas21 replied:
Implemented as part of #14

CKGrafico commented on packages/hint-compat-api/src/compat-api-css-next.ts:
```diff
@@ -0,0 +1,149 @@
+            this.testFeatureIsSupportedInBrowser(...params);
+        });
+
+        context.on('parse::css::end', async (styleParse: StyleParse) => {
```
Same as before

CKGrafico commented on packages/hint-compat-api/src/compat-api-css.ts:
```diff
@@ -0,0 +1,124 @@
+ */
+
+export default class implements IHint {
+    public static readonly meta: HintMetadata = {
```
Done with merge

CKGrafico commented on packages/hint-compat-api/src/compat-api-css.ts:
```diff
@@ -0,0 +1,124 @@
+            this.testFeatureIsSupportedInBrowser(...params);
+        });
+
+        context.on('parse::css::end', async (styleParse: StyleParse) => {
```
Same as before

CKGrafico commented on packages/hint-compat-api/src/helpers/cached-compat-features.ts:
```diff
@@ -0,0 +1,52 @@
+            return;
+        }
+
+        await cachedErrors.forEach(async (cachedFeature: CachedFeature) => {
```
True, solved

CKGrafico commented on packages/hint-compat-api/src/compat-api-css.ts:
```diff
@@ -0,0 +1,124 @@
+    public constructor(context: HintContext) {
+        this.mdnBrowsersCollection = userBrowsers.convert(context.targetedBrowsers);
+        this.compatApi = new CompatApi('css', this.mdnBrowsersCollection);
+        this.compatCSS = new CompatCSS(context, (...params) => {
```
Same as before

CKGrafico commented:
Applied @sarvaje feedback and added some comments, thanks guys!

sarvaje commented on packages/hint-compat-api/src/compat-api-css-next.ts:
```diff
@@ -0,0 +1,131 @@
+        await this.compatCSS.searchCSSFeatures(this.compatApi.compatDataApi, this.mdnBrowsersCollection, styleParse);
+    }
+
+    private async testFeatureIsSupportedInBrowser(browsersToSupport: BrowserSupportCollection, browserToSupportName: string, browserInfo: SupportStatement, featureName: string, prefix?: string, location?: ProblemLocation): Promise<void> {
```
There are few methods with a lot of parameters (like this one).

What do you think about having a type "Feature" with all the data needed?

CKGrafico replied:
We are thinking in something like
FeatureInfo (featureName: string, prefix?: string, location?: ProblemLocation)
BrowsersInfo (browsersToSupport: BrowserSupportCollection, browserToSupportName: string, browserInfo: SupportStatement)

borgitas21 replied:
Implemented as part of #14

sarvaje replied:
as part of #14?

>We are thinking in something like
FeatureInfo (featureName: string, prefix?: string, location?: ProblemLocation)
BrowsersInfo (browsersToSupport: BrowserSupportCollection, browserToSupportName: string, browserInfo: SupportStatement)

Looks good to me

borgitas21 replied:
> as part of #14?
My mistake, I meant https://github.com/CKGrafico/hint/pull/14

borgitas21 commented:
FYI: We are refactoring a bit the `compat-api-css.ts` and `compat-api-css-next.ts` classes since we have detected many duplicated code between them.

sarvaje commented on packages/hint-compat-api/src/compat-api-css-base.ts:
```diff
@@ -0,0 +1,113 @@
+ */
+
+export default abstract class BaseCompatApiCSS implements IHint {
+    public static readonly meta = meta;
```
Is the metadata common?

borgitas21 replied:
Yes, you are right!

sarvaje commented on packages/hint-compat-api/src/compat-api-css-base.ts:
```diff
@@ -0,0 +1,113 @@
+        this.compatApi = new CompatApi('css', this.mdnBrowsersCollection, isCheckingNotBroadlySupported);
+        this.compatCSS = new CompatCSS(context, this.testFeatureIsSupportedInBrowser.bind(this));
+
+        context.on('parse::end::css', async (styleParse: StyleParse) => {
```
use `this.onParseCSS(styleParse).bind(this)` instead.

sarvaje commented on packages/hint-compat-api/src/compat-api-css-base.ts:
```diff
@@ -0,0 +1,113 @@
+        const browserFeatureSupported = this.compatApi.getSupportStatementFromInfo(browser.browserInfo, feature.prefix);
+
+        if (browserFeatureSupported) {
+            this.testVersionByBrowsers(browser, browserFeatureSupported, feature);
```
`testVersionByBrowsers` is an async function, you need to wait for it to finish.
```suggestion
            await this.testVersionByBrowsers(browser, browserFeatureSupported, feature);
```

borgitas21 replied:
👍

sarvaje commented on packages/hint-compat-api/src/compat-api-css-base.ts:
```diff
@@ -0,0 +1,113 @@
+        }
+    }
+
+    private async testVersionByBrowsers(browser: BrowsersInfo, browserFeatureSupported: SimpleSupportStatement, feature: FeatureInfo) {
```
I think we should keep always as much the order as possible:
1st. Browser Info
2nd. Feature Info
3rd other Info

What do you think?

borgitas21 replied:
+1

sarvaje commented on packages/hint-compat-api/src/compat-api-css-next.ts:
```diff
@@ -0,0 +1,60 @@
+ * ------------------------------------------------------------------------------
+ */
+
+/*
```
remove commented code

sarvaje commented on packages/hint-compat-api/src/compat-api-css-next.ts:
```diff
@@ -0,0 +1,60 @@
+ */
+
+export default class extends BaseCompatApiCSS {
+    /* private userPrefixes: UserPrefixes = {}; */
```
remove commented code

borgitas21 replied:
Done

sarvaje commented on packages/hint-compat-api/src/compat-api-css-next.ts:
```diff
@@ -0,0 +1,60 @@
+        return CSSFeatureStatus.Added;
+    }
+/*
+ * private addUserUsedPrefixes(browserName: string, featureName: string): void {
```
Remove commented code

borgitas21 replied:
Done

CKGrafico commented:
@webhintio/contributors Is this ready to close? we are going start new feature

molant commented:
@CKGrafico don't wait for us to start in the new hint. Hopefully we will get this review and merged today or tomorrow.

CKGrafico commented:
Ok thanks @molant we are going to create a new PR (from this branch to a new one called `feature/hint-compat-api-html`) with @borgitas21 and request `status:work-in-progress`

borgitas21 commented:
@antross it sounds good to me grouping errors by feature. However, I'm not sure how to handle standard and prefixed reports. Could you provide me the output for the box-lines example? Thanks :)

antross commented:
> However, I'm not sure how to handle standard and prefixed reports. Could you provide me the output for the box-lines example? Thanks :)

@borgitas21 I think I see the confusion as that specific line of code didn't appear in my screenshot. The only reference to `box-lines` in my source file was `-webkit-` prefixed. All of the `box-lines` reports were generated in reference to this line of code (whether the report referred to a `-webkit-` prefix or not):

```css
    -webkit-box-lines: single;
```

Had the code contained a separate, un-prefixed reference to `box-lines`, I think a separate message for that would still be reasonable. Perhaps the "standard" reports in the screenshot were a bug?

borgitas21 commented:
After our latest changes, the amount of reports should be reduced drastically. Please let us know your feedback.

molant commented:
🎉🎉🎉

@antross can you please merge it then?

Thanks!


________________________________
From: Tony Ross <notifications@github.com>
Sent: Tuesday, December 11, 2018 14:32
To: webhintio/hint
Cc: Antón Molleda; Mention
Subject: Re: [webhintio/hint] Feature/hint compat api css (#1443)


@antross approved this pull request.

I tried the latest updates out again with our VS Code extension and the progress looks great. Thanks for all of your work putting this together!

This looks ready to merge in so we can have the foundation as part of webhint. I suspect we'll keep testing it on a growing set of real-world content before turning it on-by-default, but any issues kicked up by that process can be handled as separate bugs going forward.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/webhintio/hint/pull/1443#pullrequestreview-183936349>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AAlBghIytiYKGznW7EmR7aRDxo9gQ-qGks5u4DJ1gaJpZM4X5EZt>.

CKGrafico commented:
Amazing work everyone!
